adding katacoda tutorials

sbawaska opened the pull request:
katacoda scenarios let people try out Knative without installing anything on their local machines. The goal is add a number of basic as well as advanced scenarios that would serve as a guided tutorial for knative. This PR adds four total scenarios:
1. Getting started with Serving using kn cli
2. Getting started with Serving using yaml 
3. Getting started with Eventing with Channels
4. Getting started with Eventing with Broker

You can run these tutorials at: https://www.katacoda.com/knative-tutorials/

sbawaska commented:
<img width="1670" alt="Screen Shot 2021-01-12 at 9 50 22 PM" src="https://user-images.githubusercontent.com/1129281/104414254-d2dfd480-5524-11eb-8dcb-00f9dd26d748.png"> Once merged, I will create a PR to the website repo as well.

abrennan89 commented:
/ok-to-test

abrennan89 commented:
@sbawaska I love this, thank you for taking the time to add it.
I've added @RichieEscarez to the review from a website perspective, I'll try to take a look at the content this week.

mpetason commented on tutorials/katacoda/1-serving-intro/index.json:
```diff
@@ -0,0 +1,35 @@
+{
+  "title": "Getting Started with Knative Serving (kn cli)",
+  "description": "Introduction to Knative by installing knative serving and deploy an application (If you want to script your workflow, or really don't like yaml)",
```
```suggestion
  "description": "Introduction to Knative by installing knative serving and deploying an application (If you want to script your workflow, or really don't like yaml)",
```

mpetason commented on tutorials/katacoda/1-serving-intro/intro.md:
```diff
@@ -0,0 +1,7 @@
+## What is Knative?
+Knative brings "serverless" experience to kubernetes. It also tries to codify common patterns and best practices for running applications while hiding away the complexity of doing that on kubernetes. It does so by providing two components:
```
```suggestion
Knative brings the "serverless" experience to kubernetes. It also tries to codify common patterns and best practices for running applications while hiding the complexity of doing so on kubernetes. It does so by providing two components:
```

mpetason commented on tutorials/katacoda/2-serving-intro-yaml/intro.md:
```diff
@@ -0,0 +1,7 @@
+## What is Knative?
+Knative brings "serverless" experience to kubernetes. It also tries to codify common patterns and best practices for running applications while hiding away the complexity of doing that on kubernetes. It does so by providing two components:
```
```suggestion
Knative brings the "serverless" experience to kubernetes. It also tries to codify common patterns and best practices for running applications while hiding the complexity of doing so on kubernetes. It does so by providing two components:
```

mpetason commented on tutorials/katacoda/3-eventing-intro-channels/step3.md:
```diff
@@ -0,0 +1,115 @@
+![1toN](./assets/1toN.png)
+
+This manner of handling multiple consumers for an event, also called the fanout pattern, introduces a bunch of complex
+problems (like what if the producer crashes after delivering event to only a subset of consumers?
```
```suggestion
problems (like what if the producer crashes after delivering an event to only a subset of consumers?
```

mpetason reviewed (commented):
Initial review - I need to take some time and read through the rest again.

csantanapr commented on tutorials/katacoda/1-serving-intro/step1.md:
````diff
@@ -0,0 +1,20 @@
+
+1. Install Knative Serving's core components
+    ```
+    kubectl apply --filename https://github.com/knative/serving/releases/download/v0.19.0/serving-crds.yaml
````
harcoding the version would be hard to maintain `v0.19.0` is located in 23 places in this PR

@abrennan89 Do you know if there is a variable we can use per branch to extract the `0.19` ?

abrennan89 replied:
@markusthoemmes is probably a better person to ask

sbawaska replied:
elsewhere in the docs, the version number is referred like this: `kubectl apply --filename {{< artifact repo="serving" file="serving-crds.yaml" >}}`. That works great in the tooling used by the website, however,  these files are going to be consumed by katacoda, hence it is not possible to use that same syntax. 

How do we feel about updating these versions as part of the release process?

csantanapr replied:
Yeah that a possibility to update for each release cut, but would need to be automated with the scripts that cut a release or update api ref spec @RichieEscarez migjt have guidance on location of scripts

csantanapr replied:
I still see issues having this pointing to an old version of knative. For example we cut version 0.21

Perhaps what do you think if we don't hardcore a specific version and instead we use the nightly URL instead?

csantanapr replied:
Oh wait this is running on Linux not users computer, I proposed you use bash/sh for example cusinf curl to find out the latest cut version and stored that in variable then use the variable when installing with kubectl URL

abrennan89 commented:
Folks this is not something I will have bandwidth to review any time soon, and I'm not sure that Katacoda tutorials content reviews should really be part of Docs WG scope anyway, since they're kind of their own thing outside of our usual written markdown content.

@omerbensaadon is this something you can look at as part of the UX WG instead maybe?

evankanderson commented:
/assign @omerbensaadon

abrennan89 commented:
/retest

csantanapr reviewed (changes requested):
This PR is a good idea, but right now is block until it get's updated to remove the hardcoded version of v0.19. It needs to be a bash script that fetches what's the latest versioned released from github released, and uses that version.

abrennan89 commented:
/hold

Blocked as per Carlos's comment

csantanapr commented on tutorials/katacoda/1-serving-intro/step1.md:
````diff
@@ -0,0 +1,20 @@
+    ```{{execute}}
+1. Install contour as the networking layer. (Knative also supports Courier, Gloo, Istio and Kourier as options)
+    ```
+    kubectl apply --filename https://github.com/knative/net-contour/releases/download/v0.19.0/contour.yaml
````
I would prefer to use kourier instead of contour, less required resources

sbawaska commented:
> This PR is a good idea, but right now is block until it get's updated to remove the hardcoded version of v0.19. It needs to be a bash script that fetches what's the latest versioned released from github released, and uses that version.

Hi @csantanapr , the hardcoded versions have been removed. Please review.

omerbensaadon commented:
missed the assignment here but it looks like @csantanapr had me covered 🥇

omerbensaadon reviewed (commented):
Would be a great addition to the site! 

Have we narrowed in on where this would live on the site?

sbawaska commented:
> Would be a great addition to the site!
> 
> Have we narrowed in on where this would live on the site?

I was thinking this: https://github.com/knative/website/pull/246

mpetason commented on tutorials/katacoda/3-eventing-intro-channels/step4.md:
````diff
@@ -0,0 +1,101 @@
+EOF
+```{{execute}}
+
+Now, let's create the Sequence to execute the service "first" and then pass it's result to the service "event-display-chain"
````
```suggestion
Now, let's create the Sequence to execute the service "first" and then pass its result to the service "event-display-chain"
```

mpetason commented on tutorials/katacoda/4-eventing-intro-broker/step2.md:
````diff
@@ -0,0 +1,111 @@
+EOF
+```{{execute}}
+
+Now We will create consumers that will simply log the event. (We will only create two consumers, rest all will be very similar):
````
```suggestion
Now we will create consumers that will simply log the event. (We will only create two consumers, the rest will be very similar):
```

mpetason commented on tutorials/katacoda/4-eventing-intro-broker/step2.md:
```diff
@@ -0,0 +1,111 @@
+## Use Case
+The broker and trigger model is useful for complex event delivery topologies like N:M:Z, i.e. there are a multitude of Sources sending events, Functions
+consuming/transforming which are then processed by even more functions and so on. It can get a bit unwieldy to keep track of which Channel is having which events. Also, sometimes you might only want to
+consume only some types of event. You would have to receive all the events, and throw out the ones you’re not interested in. To make these kinds of interactions easier and allow the
```
```suggestion
consuming/transforming which are then processed by even more functions and so on. It can get a bit unwieldy to keep track of which Channel is having which events. Also, sometimes you might only want to
consume specific types of events. You would have to receive all the events, and throw out the ones you’re not interested in. To make these kinds of interactions easier and allow the
```

mpetason reviewed (commented):
This looks great! I added a few ideas that may read a little bit better plus a few small updates. Great work.

sbawaska commented:
> This looks great! I added a few ideas that may read a little bit better plus a few small updates. Great work.

Thanks for the review @mpetason. I have incorporated your feedback.

csantanapr commented on tutorials/katacoda/1-serving-intro/step2.md:
````diff
@@ -0,0 +1,26 @@
+Knative defines a `service.serving.knative.dev` CRD to control the lifecycle of the application (not to be confused with kubernetes service). We will use the `kn` cli to create the Knative service:
+
+```
+kn service create demo --image gcr.io/knative-samples/helloworld-go --env TARGET=blue
````
```suggestion
kn service create demo --image gcr.io/knative-samples/helloworld-go --env TARGET=blue --autoscale-window 15s
```

what do you think of specify a 15 seconds, this way the user doesn't have to seat there waiting for 60 seconds

csantanapr commented on tutorials/katacoda/1-serving-intro/step2.md:
```diff
@@ -0,0 +1,26 @@
+
+We will now deploy the application by specifying the image location and the `TARGET` env variable.
+
+Knative defines a `service.serving.knative.dev` CRD to control the lifecycle of the application (not to be confused with kubernetes service). We will use the `kn` cli to create the Knative service:
```
Do you want to mention something that this step will take up to a minute, at least when I try it is taking about 1 minute?
Is to avoid the user thinking that something is wrong.

csantanapr replied:
```
$ kn service create demo --image gcr.io/knative-samples/helloworld-go --env TARGET=blue
Creating service 'demo' in namespace 'default':

  0.155s The Route is still working to reflect the latest desired specification.
  0.218s Configuration "demo" is waiting for a Revision to become ready.
 50.677s ...
 50.816s Ingress has not yet been reconciled.
 50.874s Waiting for Envoys to receive Endpoints data.
 51.110s Waiting for load balancer to be ready
 51.329s Ready to serve.
 ```

sbawaska replied:
added a note.

csantanapr commented on tutorials/katacoda/1-serving-intro/step2.md:
````diff
@@ -0,0 +1,26 @@
+```{{execute}}
+Then invoke the application using curl:
+```
+curl http://$MINIKUBE_IP:$INGRESS_PORT/ -H 'Host: demo.default.example.com'
````
It would be great to configure the minikube to use a local domain name such as `http://demo.default.127.0.0.1.nip.io`
Or a local IP `10.` when I use minikube I use `minikube tunnel` to get a local tunnel on an IP address for a load balancer IP.

sbawaska replied:
I think we can fix that in a follow on PR.

csantanapr commented on tutorials/katacoda/1-serving-intro/step2.md:
````diff
@@ -0,0 +1,26 @@
+```{{execute T1}}
+
+### Scale down to zero
+You can run `watch kubectl get pods`{{execute T2}} in a new Terminal tab to see a pod created to serve the requests. Knative will scale this pod down to zero if there are no incoming requests for 60 seconds by default.
````
When I click `watch kubectl get pods` It opens the terminal 2 in the right, side but nothing shows up.

Need to tell the user what to do to make the command output to show up
![image](https://user-images.githubusercontent.com/1094878/112038693-8f53a900-8b19-11eb-8373-f58c6999c58d.png)

csantanapr replied:
I ad to click twice, maybe one is to focus the T2 terminal and the second one is to run the command, not sure

sbawaska replied:
Yes, I have seen it sometimes too. It probably is a bug in katacoda, I added a note to click twice.

csantanapr commented on tutorials/katacoda/1-serving-intro/step3.md:
```diff
@@ -0,0 +1,28 @@
+## Blue/Green deploy
+The Knative `service` resource creates additional resources "route, configuration and revision" to manage the lifecycle of the application.
```
This section might be confusing for new users.

- Title mentions "Blue/Green" but there is not rollout from current version (blue) to the new version (green)
- The traffic is left at 50/50 traffic split, the new version never goes to 100%
- A tag is used in the example but the tag is never use, the tag is for the purpose of unique URL, but it was never use
- There is a command that feels like brack magic `kn revision list -o 'jsonpath={.items[0].metadata.name}`

Here is a [tutorial](https://github.com/csantanapr/knative-tekton#33-knative-service-traffic-splitting) with a flow that is more appropriate. 

We could use a revision name called it revision `blue`, and a revision `green` green being latest.
You can show that green version gets 50% traffic and blue 50%, and then show blue takes 0% and green 100%

sbawaska replied:
This was super helpful. I have incorporated this. Thanks!

csantanapr commented on tutorials/katacoda/2-serving-intro-yaml/step3.md:
```diff
@@ -0,0 +1,68 @@
+## Blue/Green deploy
+The Knative `service` resource creates additional resources "route, configuration and revision" to manage the lifecycle of the application.
```
I have similar feedback I gave back in [the cli section](https://github.com/knative/docs/pull/3142/files#r599149635)

sbawaska replied:
fixed!

csantanapr commented on tutorials/katacoda/3-eventing-intro-channels/step2.md:
```diff
@@ -0,0 +1,93 @@
+### 1:1 Event Delivery
+The most straightforward use case is that whenever events are produced, you want some code to handle that event.
+
+![1to1](./assets/1to1.png)
```
Really like the diagram pictures ❤️ 

Wondering if we can make diagrams for serving tutorial

sbawaska replied:
Thanks! Will consider in a follow on PR

csantanapr commented on tutorials/katacoda/3-eventing-intro-channels/step2.md:
````diff
@@ -0,0 +1,93 @@
+kubectl wait --for=condition=ready pod -l serving.knative.dev/service=event-display
+# see the logs
+kubectl logs -l serving.knative.dev/service=event-display -c user-container --since=10m --tail=50
+```{{execute}}
````
I saw problems with this execute, I had to click twice to make finally show up

sbawaska replied:
problem with execute might be a browser thing? I don't see issues with chrome. What browser are you using?

csantanapr commented on tutorials/katacoda/3-eventing-intro-channels/step2.md:
````diff
@@ -0,0 +1,93 @@
+
+```
+# it is likely that is pod is still being created after scaling down to zero
+kubectl wait --for=condition=ready pod -l serving.knative.dev/service=event-display
````
I like the approach of the wait for condition and then show the logs

sbawaska replied:
Thanks!

csantanapr commented on tutorials/katacoda/3-eventing-intro-channels/step2.md:
```diff
@@ -0,0 +1,93 @@
+# it is likely that is pod is still being created after scaling down to zero
+kubectl wait --for=condition=ready pod -l serving.knative.dev/service=event-display
+# see the logs
+kubectl logs -l serving.knative.dev/service=event-display -c user-container --since=10m --tail=50
```
The logs don't show up on first run I click.
Maybe we need to put some type of sleep.

I had to run it twice to see the logs

sbawaska replied:
The default timeout was 30 seconds, bumped it to 90.

csantanapr commented on tutorials/katacoda/3-eventing-intro-channels/step4.md:
````diff
@@ -0,0 +1,101 @@
+
+```
+cat <<EOF | kubectl create -f -
+apiVersion: flows.knative.dev/v1beta1
````
Update all the `v1beta1` to `v1`

sbawaska replied:
done!

csantanapr commented on tutorials/katacoda/3-eventing-intro-channels/step4.md:
````diff
@@ -0,0 +1,101 @@
+kubectl wait --for=condition=ready pod -l serving.knative.dev/service=event-display-chain
+# see the logs
+kubectl -n default logs -l serving.knative.dev/service=event-display-chain -c user-container --since=10m --tail=50
+```{{execute}}
````
I had to run it twice to see the output of the logs

there was a timeout error also

```
pingsource.sources.knative.dev/ping-source-sequence created
$ # it is likely that is pod is still being created after scaling down to zero
$ kubectl wait --for=condition=ready pod -l serving.knative.dev/service=event-display-chain
pod/event-display-chain-00001-deployment-566755999b-g4xqg condition met
$ # see the logs
$ kubectl -n default logs -l serving.knative.dev/service=event-display-chain -c user-container --since=10m --tail=50
$ # it is likely that is pod is still being created after scaling down to zero
$ kubectl wait --for=condition=ready pod -l serving.knative.dev/service=event-display-chain
error: timed out waiting for the condition on pods/event-display-chain-00001-deployment-566755999b-g4xqg
$ # see the logs
$ kubectl -n default logs -l serving.knative.dev/service=event-display-chain -c user-container --since=10m --tail=50
☁️  cloudevents.Event
Validation: valid
Context Attributes,
  specversion: 1.0
  type: dev.knative.sources.ping
  source: /apis/v1/namespaces/default/pingsources/ping-source-sequence
  id: d65624c9-70dd-4bbd-8ec0-d1bfbfbb39dd
  time: 2021-03-23T01:59:00.103779523Z
  datacontenttype: application/json
Data,
  {
    "id": 0,
    "message": "Hello Sequence! - Updating this message a little..."
  }
```

sbawaska replied:
The default timeout was 30 seconds, bumped it to 90.

csantanapr commented on tutorials/katacoda/4-eventing-intro-broker/step2.md:
```diff
@@ -0,0 +1,111 @@
+
+Let us now create Triggers that will send only the events that the consumer is interested in:
+
+For `consumer2`, we create a filter for Red events only
```
Is kind odd to start with consumer2 and then consumer1, I would expect you do them in order starting with something named consumer1

sbawaska replied:
Good point. I have consumer1 creating two triggers, hence wanted to start with consumer2 first. Is it ok to fix this later?

csantanapr commented on tutorials/katacoda/4-eventing-intro-broker/step3.md:
````diff
@@ -0,0 +1,77 @@
+kubectl wait pods/post-event --for=condition=ContainersReady
+```{{execute}}
+
+Let us publish some [cloudevents](https://cloudevents.io/). The command below adds some cloud-events specific headers to the curl command. Publish red events using the command:
````
add break line before last sentence.

Also code quotes for the colors `red` `blue` `orange`

sbawaska replied:
done!

csantanapr commented on tutorials/katacoda/4-eventing-intro-broker/step3.md:
````diff
@@ -0,0 +1,77 @@
+# see the logs
+kubectl logs -l serving.knative.dev/service=consumer1 -c user-container --since=10m --tail=50
+kubectl logs -l serving.knative.dev/service=consumer2 -c user-container --since=10m --tail=50
+```{{execute}}
````
This is not working for `consumer2`

```
$ # see the logs
$ kubectl logs -l serving.knative.dev/service=consumer1 -c user-container --since=10m --tail=50
☁️  cloudevents.Event
Validation: valid
Context Attributes,
  specversion: 1.0
  type: orange
  source: kata-tutorial
  id: test
  datacontenttype: application/json
Extensions,
  knativearrivaltime: 2021-03-23T02:14:43.808385659Z
Data,
  {
    "msg": "The event is orange!"
  }
$ kubectl logs -l serving.knative.dev/service=consumer2 -c user-container --since=10m --tail=50
$ # it is likely that is pod is still being created after scaling down to zero
$ kubectl wait --for=condition=ready pod -l serving.knative.dev/service=consumer1
pod/consumer1-00001-deployment-5b5b8c849d-cb8gh condition met
$ kubectl wait --for=condition=ready pod -l serving.knative.dev/service=consumer2
error: no matching resources found
$ # see the logs
$ kubectl logs -l serving.knative.dev/service=consumer1 -c user-container --since=10m --tail=50
☁️  cloudevents.Event
Validation: valid
Context Attributes,
  specversion: 1.0
  type: orange
  source: kata-tutorial
  id: test
  datacontenttype: application/json
Extensions,
  knativearrivaltime: 2021-03-23T02:14:43.808385659Z
Data,
  {
    "msg": "The event is orange!"
  }
$ kubectl logs -l serving.knative.dev/service=consumer2 -c user-container --since=10m --tail=50
```

sbawaska replied:
The default timeout was 30 seconds, bumped it to 90.

csantanapr commented:
/hold

csantanapr commented:
@sbawaska I did a review. 
Please take a at the comments and also the errors I saw while trying the tutorial

sbawaska commented on tutorials/katacoda/3-eventing-intro-channels/step2.md:
````diff
@@ -0,0 +1,93 @@
+kubectl logs -l serving.knative.dev/service=event-display -c user-container --since=10m --tail=50
+```{{execute}}
+
+### N:1 Event Delivery
````
Thanks!

sbawaska commented:
> @sbawaska I did a review.
> Please take a at the comments and also the errors I saw while trying the tutorial

Thanks for the thorough review @csantanapr! I have incorporated most of it. please review. Thanks!

csantanapr commented:
Hi @sbawaska 

You mentioned that you preferred to fix some items in a subsequent PR, would you mind collecting a these in an issue so you or others contributors could improve once this is merge?

I would like to contribute to it as part of new initiatives in the UX WG
cc @omerbensaadon

omerbensaadon commented:
@sbawaska we are planning to test this in User Interviews beginning early next week! 

Not a deadline, we can test the basic premise (Getting Started w/ katacoda) with what we have, just thought I would let you know.

csantanapr commented:
/lgtm
/approve

csantanapr commented:
/unhold
